packages/core-contracts/contracts/token/ERC721.sol

```diff
@@ -0,0 +1,243 @@
+        return _erc721Store().symbol;
+    }
+
+    function tokenURI(uint256 tokenId) external view virtual override returns (string memory) {
+        if (!_exists(tokenId)) {
+            revert TokenDoesNotExist(tokenId);
+        }
+
+        string memory baseURI = _erc721Store().baseTokenURI;
+        return bytes(baseURI).length > 0 ? string(abi.encodePacked(baseURI, _toString(tokenId))) : "";
+    }
+
+    function approve(address to, uint256 tokenId) public virtual override {
+        ERC721Store storage store = _erc721Store();
+        address holder = store.ownerOf[tokenId];
+
+        if (to == holder) {
+            revert CannotApproveToHolder(to);
+        }
+
+        if (msg.sender != holder && !isApprovedForAll(holder, msg.sender)) {
+            revert Unauthorized(msg.sender);
+        }
+
+        _approve(to, tokenId);
+    }
+
+    function getApproved(uint256 tokenId) public view virtual override returns (address) {
+        if (!_exists(tokenId)) {
+            revert TokenDoesNotExist(tokenId);
+        }
+
+        return _erc721Store().tokenApprovals[tokenId];
+    }
+
+    function setApprovalForAll(address operator, bool approved) public virtual override {
+        if (msg.sender == operator) {
+            revert CannotApproveToCaller(operator);
+        }
+
+        _erc721Store().operatorApprovals[msg.sender][operator] = approved;
+        emit ApprovalForAll(msg.sender, operator, approved);
+    }
+
+    function isApprovedForAll(address holder, address operator) public view virtual override returns (bool) {
+        return _erc721Store().operatorApprovals[holder][operator];
+    }
+
+    function transferFrom(
+        address from,
+        address to,
+        uint256 tokenId
+    ) public virtual override {
+        if (!_isApprovedOrOwner(msg.sender, tokenId)) {
+            revert Unauthorized(msg.sender);
+        }
+
+        _transfer(from, to, tokenId);
+    }
+
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 tokenId
+    ) public virtual override {
+        safeTransferFrom(from, to, tokenId, "");
+    }
+
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 tokenId,
+        bytes memory data
+    ) public virtual override {
+        if (!_isApprovedOrOwner(msg.sender, tokenId)) {
+            revert Unauthorized(msg.sender);
+        }
+
+        _transfer(from, to, tokenId);
+        if (!_checkOnERC721Received(from, to, tokenId, data)) {
+            revert InvalidTransferRecipient(to);
+        }
+    }
+
+    function _exists(uint256 tokenId) internal view virtual returns (bool) {
+        return _erc721Store().ownerOf[tokenId] != address(0);
+    }
+
+    function _isApprovedOrOwner(address spender, uint256 tokenId) internal view virtual returns (bool) {
+        address holder = ownerOf(tokenId);
+
+        // Not checking tokenId existence since it is checked in ownerOf() and getApproved()
```

Comment:
What about execution flows that don't enter through the `ownerOf` and `getApproved` interfaces. E.g. `transferFrom` and `safeTransferFrom`?

In those cases, the `_exists` call is not made in parallel.

Reply:
Ah nevermind! It's checked in ownerOf in line 152